### DISCUSSION TITLE
Rules block restart

### DISCUSSION THREAD

**Discussion by tjareson** [arendst/Tasmota#14259] [Q&A] [answered]
Hello,
I have 2 rules defined:
rule1:
on var4#state=1 do backlog delay %var2%; power2 off; var4 0 endon on var4#state=0 do backlog delay %var3%; power2 %var1%; var4 1 endon
rule2:
rule2 on Power1#Boot do backlog var1 1; var2 50; var3 200; var4 1 endon

Rule1 is a very slow "PWM" for a valve which opens by getting heated up connected via a triac controlled by power2. 
Rule2 takes care that the variables getting initialized, when the device is connected to power. The variables are changed via MQTT commands by a central controlling software. Everything work as designed, except one thing:
If rule1 is running( so var4 is changed at least once after start, usually by rule2), it is not possible to restart tasmota. 
Only if I set rule1 to off it restarts. I'm using 9.3.1 but have the same with other tasmota versions.
Does anyone have an explanation for that behaviour? Is restart blocked as long as there are commands in the backlog?

**Comment by sfromis** [comment, marked as answer]
Using the Delay command is not reliable for timing in rules, as it can be cut short by things like commands. RuleTimer would be better, or maybe look into using PulseTime, where Tasmota will automatically turn off again after expiry of the timer; would simplify the timing loop.

Without details of what you mean, including console log, "not possible to restart" does not convey clues about what may be happening.

**Reply by tjareson** [reply]
Ruletimer would require a binary different to the precompiled ones, which I'm trying to avoid to be able to update easier, in case it is necessary. 
The Pulsetime has the issue that I don't know how to create the pauses in between then. So like 200 seconds ON then 50 seconds OFF for example. That doesn't look possible with Pulsetime alone. Also I'm not sure if changed pulse times are getting written into the eeprom. As this value get's changed frequently via mqtt, this would wear out the eeprom unnecessarily over time. 

The timing doesn't have to be overly precise and delay in backlog is also part of the documentation of explaning the possibilities of the backlog function itself. So I wasn't expecting to have done something super uncommon here. 

Related to the console log: if I press the restart button in the menu, nothing shows up there at all. It just continues to run the rules. And the device just didn't restart, as uptime didn't reset to zero. (device is a sonoff basic by the way)

If I try to do a restart 1 on the console this happens:
13:26:57.990 RSL: stat/tasmota_579CDF/POWER2 = OFF
13:26:58.208 RSL: stat/tasmota_579CDF/RESULT = {"Var4":"0"}
13:26:58.237 RUL: VAR4#STATE=0 performs "backlog delay 200; power2 1; var4 1"
13:26:58.246 RSL: stat/tasmota_579CDF/RESULT = {"Delay":200}
13:27:02.913 CMD: restart 1
13:27:02.919 RSL: stat/tasmota_579CDF/RESULT = {"Restart":"Restarting"}
13:27:03.161 RSL: stat/tasmota_579CDF/RESULT = {"POWER2":"ON"}
13:27:03.164 RSL: stat/tasmota_579CDF/POWER2 = ON
13:27:03.412 RSL: stat/tasmota_579CDF/RESULT = {"Var4":"1"}
13:27:03.434 RUL: VAR4#STATE=1 performs "backlog delay 50; power2 off; var4 0"
13:27:03.461 RSL: stat/tasmota_579CDF/RESULT = {"Delay":50}
13:27:08.480 RSL: stat/tasmota_579CDF/RESULT = {"POWER2":"OFF"}
13:27:08.483 RSL: stat/tasmota_579CDF/POWER2 = OFF
13:27:08.731 RSL: stat/tasmota_579CDF/RESULT = {"Var4":"0"}
13:27:08.755 RUL: VAR4#STATE=0 performs "backlog delay 200; power2 1; var4 1"
13:27:08.779 RSL: stat/tasmota_579CDF/RESULT = {"Delay":200}

It just continues, as nothing has happened. Interesting here is: as soon as I stop the rule with rule1 0 it reboots. So it looks like it just waits with the restart until the rule and/or backlog commands are finished. Which is somehow a bit unexpected behaviour. 
Another consequence is that any change of configuration via web doesn't work, if it says save&restarting now. It never restarts. In case of power cycling it, the changes are also not saved. As soon as I stop rule1 it works again. 

What makes it more confusing is that restart 99 (so restart w/o saving any configuration) works. So it looks to me like as it is refusing to save the configuration data, while rules/backlog commands are running, hence it waits with the restart.

savedata is off by the way - which probably would have been an idea of how the rule which is constantly changing the output could have interfered the saving process of a regular restart.

**Reply by sfromis** [reply]
No, RuleTimer is a standard feature in precompiled binaries. Using PulseTime makes the RuleTimer Setup simpler, as you do not have to have extra timing to yourself switch off again after the set period. But indeed PulseTime is a setup saved to flash, unless you hinder that. However, with you doing Power on/off commands in the loop, this will already cause flash writes (again unless hindered), so this is similar.

As for the Delay command, I just wanted you to be aware that this is unreliable, especially for more than brief intervals. 

It does indeed look like the ongoing tick-tock loop means that it will not restart while a rule is active. This is just another good reason for using RuleTimer instead, as you will not have a rule in execution all the time.

**Reply by tjareson** [reply]
Ok, cool, thanks! Somehow I had in mind ruletimer was an option for self compiling only. I've just adjusted my delays to use 2 ruletimers instead, tweaked my node red setup to send values in seconds and it works including restart. :-)
